Release/src/http/client/http_client_winhttp.cpp

```diff
@@ -1100,136 +1028,88 @@ class winhttp_client : public _http_client_communicator
-						}
-
-						if (!pinned)
-						{
-                            // Report back to the client with the failed certs when cert-pinning has failed.
-                            p_request_context->m_request.invoke_rejected_certificate_chain_callback(certificatesInfo);
-							// client refused all the certificates, track the failure and cancel the request.
-							p_request_context->m_certificate_pinning_failed = true;
-							p_request_context->cleanup();
-							return;
-						}
-					}
-					else
-					{
-						// not able to extract the leaf certificate, ask the pinning callback what to do in this case by passing in null values
-						if (!p_request_context->m_http_client->client_config().invoke_pinning_callback(url, U("")))
-						{
-							// connection not allowed to continue with no certificates available, callback with the failure and cancel the request.
-                            p_request_context->m_request.invoke_rejected_certificate_chain_callback( {} );
-							p_request_context->m_certificate_pinning_failed = true;
-							p_request_context->cleanup();
-							return;
-						}
-					}
-
-					break;
-				}
+            {
+                // get actual URL which might be different from the original one due to redirection etc.
+                DWORD urlSize{ 0 };
+                WinHttpQueryOption(hRequestHandle, WINHTTP_OPTION_URL, NULL, &urlSize);
+                auto urlwchar = new WCHAR[urlSize / sizeof(WCHAR)];
+
+                WinHttpQueryOption(hRequestHandle, WINHTTP_OPTION_URL, (void*)urlwchar, &urlSize);
+                utility::string_t url(urlwchar);
+
+                delete[] urlwchar;
+
+                // obtain leaf cert based on which we will be able to build the certificate chain
+                PCCERT_CONTEXT pCert{ nullptr };
+                DWORD dwSize = sizeof(pCert);
+
+                WinHttpQueryOption(hRequestHandle, WINHTTP_OPTION_SERVER_CERT_CONTEXT, &pCert, &dwSize);
+
+                std::vector<std::pair<const unsigned char*, int>> cert_chain;
+
+                if (pCert)
+                {
+                    CERT_ENHKEY_USAGE keyUsage = {};
+                    keyUsage.cUsageIdentifier = 0;
+                    keyUsage.rgpszUsageIdentifier = NULL;
+
+                    CERT_USAGE_MATCH certUsage = {};
+                    certUsage.dwType = USAGE_MATCH_TYPE_AND;
+                    certUsage.Usage = keyUsage;
+
+                    CERT_CHAIN_PARA chainPara = {};
+                    chainPara.cbSize = sizeof(CERT_CHAIN_PARA);
+                    chainPara.RequestedUsage = certUsage;
+
+                    PCCERT_CHAIN_CONTEXT pChainContext = {};
+
+                    // build the certificate chain relying on the actual intermediate certs returned as part of the TLS session
+                    // disable any network operations to fetch certificates
+                    auto validChain = CertGetCertificateChain(NULL, pCert, NULL, pCert->hCertStore, &chainPara,
```

Comment:
I think we're missing a call to `CertFreeCertificateChain`

Ref: https://msdn.microsoft.com/en-us/library/windows/desktop/aa376078(v=vs.85).aspx